src/vulkan/vertex_buffer.cc

```diff
@@ -23,70 +23,78 @@
+  if (rate == InputRate::kVertex)
+    return VK_VERTEX_INPUT_RATE_VERTEX;
+  else
+    return VK_VERTEX_INPUT_RATE_INSTANCE;
+}
+
+}  // namespace
 
 VertexBuffer::VertexBuffer(Device* device) : device_(device) {}
 
 VertexBuffer::~VertexBuffer() = default;
 
-void VertexBuffer::SetData(uint8_t location, Buffer* buffer) {
+void VertexBuffer::SetData(uint8_t location, Buffer* buffer, InputRate rate) {
   auto format = buffer->GetFormat();
-
+  const uint32_t binding = static_cast<uint32_t>(vertex_attr_desc_.size());
   vertex_attr_desc_.emplace_back();
-  // TODO(jaebaek): Support multiple binding
-  vertex_attr_desc_.back().binding = 0;
+  vertex_attr_desc_.back().binding = binding;
   vertex_attr_desc_.back().location = location;
-  vertex_attr_desc_.back().offset = stride_in_bytes_;
+  vertex_attr_desc_.back().offset = 0u;
   vertex_attr_desc_.back().format = device_->GetVkFormat(*format);
 
-  stride_in_bytes_ += format->SizeInBytes();
-  data_.push_back(buffer);
-}
-
-Result VertexBuffer::FillVertexBufferWithData(CommandBuffer* command) {
-  // Send vertex data from host to device.
-  uint8_t* ptr_in_stride_begin =
-      static_cast<uint8_t*>(transfer_buffer_->HostAccessibleMemoryPtr());
-  for (uint32_t i = 0; i < GetVertexCount(); ++i) {
-    uint8_t* ptr = ptr_in_stride_begin;
-    for (uint32_t j = 0; j < data_.size(); ++j) {
-      size_t bytes = data_[j]->GetFormat()->SizeInBytes();
-      std::memcpy(ptr, data_[j]->GetValues<uint8_t>() + (i * bytes), bytes);
-      ptr += bytes;
-    }
-    ptr_in_stride_begin += Get4BytesAlignedStride();
-  }
+  vertex_binding_desc_.emplace_back();
+  vertex_binding_desc_.back().binding = binding;
+  vertex_binding_desc_.back().stride = format->SizeInBytes();
+  vertex_binding_desc_.back().inputRate = GetVkInputRate(rate);
 
-  transfer_buffer_->CopyToDevice(command);
-  return {};
+  data_.push_back(buffer);
 }
 
 void VertexBuffer::BindToCommandBuffer(CommandBuffer* command) {
-  const VkDeviceSize offset = 0;
-  const VkBuffer buffer = transfer_buffer_->GetVkBuffer();
-  // TODO(jaebaek): Support multiple binding
-  device_->GetPtrs()->vkCmdBindVertexBuffers(command->GetVkCommandBuffer(), 0,
-                                             1, &buffer, &offset);
+  std::vector<VkBuffer> buffers;
+  std::vector<VkDeviceSize> offsets;
+
+  for (const auto& buf : transfer_buffers_) {
+    buffers.push_back(buf->GetVkBuffer());
+    offsets.push_back(0);
+  }
+  device_->GetPtrs()->vkCmdBindVertexBuffers(
+      command->GetVkCommandBuffer(), 0,
+      static_cast<uint32_t>(transfer_buffers_.size()), buffers.data(),
+      offsets.data());
 }
 
 Result VertexBuffer::SendVertexData(CommandBuffer* command) {
   if (!is_vertex_data_pending_)
     return Result("Vulkan::Vertices data was already sent");
 
-  const uint32_t n_vertices = GetVertexCount();
-  if (n_vertices == 0)
-    return Result("Vulkan::Data for VertexBuffer is empty");
+  // Non-empty transfer_buffers_ means we are in a unit test.
```

Comment:
Why does this mean unit test? We shouldn't have code in here which is unit test specific.

Reply:
I totally agree. The same structure was there before and I just added comment to clarify there was unit test specific code present.

I think it's trying to avoid using a device object which is NULL when in unit test. We had a similar issue before and we used a dummy object in unit test instead. I'll try to use similar solution here.

Reply:
Managed to add a dummy device class and get rid of test specific code in vertex buffer class.